### ISSUE TITLE
RedundantVisibilityModifierRule false-positives when `explicitApi()` is defined in the `kotlin{}` extension 

### ISSUE THREAD

**Issue by eygraber** [detekt/detekt#3874]
I have `explicitApi()` enabled for my module:

```
kotlin {
  explicitApi()
}
```

but I still get warned about `RedundantVisibilityModifierRule` for public functions and classes.

## Your Environment
* Version of detekt used: 1.17.1
* Version of Kotlin used: 1.5.10
* Version of Gradle used (if applicable): 7.1

**Comment by cortinico** [comment]
Could you provide a small reproducer? I'm not entirely sure what could be the root cause here

**Comment by eygraber** [comment]
It was happening in this project https://github.com/eygraber/ejson-kotlin

The rule is currently disabled, but if you enable it, it should repro.

**Comment by cortinico** [comment]
I've done a quick investigation here.

The problem is that the Kotlin compilation invoked by Detekt is running with different flags. The Explicit API flag is not passed over to detekt, therefore is essentially ignored.

This will be resolved once we migrate to https://github.com/detekt/detekt-compiler-plugin/ as the compiler flags will be the same.

I'm unsure how we want to fix this, but I have the gut feeling that the check for `isExplicitApiModeActive` we introduced in #3358 is not effective (should we remove it @schalkms ?).

**Comment by schalkms** [comment]
Regarding the `explicitApi` modifier, this issue was fixed in a subsequent PR.

**Comment by cortinico** [comment]
> Regarding the `explicitApi` modifier, this issue was fixed in a subsequent PR.

Do you happen to recall which one?

**Comment by schalkms** [comment]
> Do you happen to recall which one?

#3358

**Comment by cortinico** [comment]
> I'm unsure how we want to fix this, but I have the gut feeling that the check for `isExplicitApiModeActive` we introduced in #3358 is not effective (should we remove it @schalkms ?).

Yup so I believe the fix we introduced in #3358 is ineffective.

**Comment by schalkms** [comment]
> Yup so I believe the fix we introduced in #3358 is ineffective.

I had to fiddle around with very low level internal compiler APIs. So maybe something changed there? It's very hard to test this E2E with automated test cases.

**Comment by schalkms** [comment]
It could also be the case that the setting `explicitApi()` isn't picked up in each and every case depending on a variety of possible options to configure this setting.

**Comment by cortinico** [comment]
> It could also be the case that the setting `explicitApi()` isn't picked up in each and every case depending on a variety of possible options to configure this setting.

I believe that's the case. Specifically the compiler flag is not passed over so is not present when Detekt is running 🤔

**Comment by cortinico** [comment]
I believe this could get fixed by https://github.com/detekt/detekt/pull/4106

**Comment by mbayerPK** [comment]
any updates on this issue? I see that something was discussed in https://github.com/detekt/detekt/pull/4106 which was closed and then was brought up again in https://github.com/detekt/detekt/issues/4465.

**Comment by cortinico** [comment]
> I see that something was discussed in #4106 which was closed and then was brought up again in #4465.

Please follow up on #4465 directly 👍

**Comment by 3flex** [comment]
Support for this is being added in https://github.com/detekt/detekt/pull/7418, but the rule needs to be updated to take advantage of this.

Use one of the following:
```kotlin
compilerResources?.languageVersionSettings?.explicitApiEnabled
compilerResources?.languageVersionSettings?.getFlag(AnalysisFlags.explicitApiMode)
```

The first is suitable if the rule only cares if explicit API mode is in strict mode or not. The second can be used to get the specific value that's been set (strict, warning or disabled).

**Comment by atulgpt** [comment]
I think logic and code are already there in the rule `RedundantVisibilityModifier` at https://github.com/detekt/detekt/blob/8b1c0c69d16d39244f5cb164f02666aab3231853/detekt-rules-style/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/RedundantVisibilityModifier.kt#L60

In the PR https://github.com/detekt/detekt/pull/7429 I have also considered a Warning for the check `isExplicitApiModeActive` as if someone sets the API to `WARNING` then they will not want detekt warning on the specifying the visibility modifier

**Comment by 3flex** [comment]
> I think logic and code are already there

Oh you're right, I guess it just never worked? Proper tests should be added.

**Comment by 3flex** [comment]
Closing though as the issue is fixed by #7418
